test/hotspot/jtreg/compiler/lib/template_framework/Template.java

```diff
@@ -0,0 +1,700 @@
+        Renderer.getCurrent().addHashtagReplacement(key, value);
+        return function.apply(value);
+    }
+
+    /**
+     * Default amount of fuel for Template rendering. It guides the nesting depth of Templates.
+     */
+    public final static float DEFAULT_FUEL = 100.0f;
+
+    /**
+     * The default amount of fuel spent per Template. It is subtracted from the current {@link #fuel} at every
+     * nesting level, and once the {@link #fuel} reaches zero, the nesting is supposed to terminate.
+     */
+    public final static float DEFAULT_FUEL_COST = 10.0f;
+
+    /**
+     * The current remaining fuel for nested Templates. Every level of Template nesting
+     * subtracts a certain amount of fuel, and when it reaches zero, Templates are supposed to
+     * stop nesting, if possible. This is not a hard rule, but a guide, and a mechanism to ensure
+     * termination in recursive Template instantiations.
+     *
+     * <p>
+     * Example of a recursive Template, which checks the remaining {@link #fuel} at every level,
+     * and terminates if it reaches zero. It also demonstrates the use of {@link TemplateBinding} for
+     * the recursive use of Templates. We {@link Template.OneArgs#render} with {@code 30} total fuel, and spend {@code 5} fuel at each recursion level.
+     * {@snippet lang=java :
+     * var binding = new TemplateBinding<Template.OneArgs<Integer>>();
+     * var template = Template.make("depth", (Integer depth) -> body(
+     *     setFuelCost(5.0f),
+     *     let("fuel", fuel()),
+     *     """
+     *     System.out.println("Currently at depth #depth with fuel #fuel");
+     *     """,
+     *     (fuel() > 0) ? binding.get().asToken(depth + 1)
+     *                    "// terminate\n"
+     * ));
+     * binding.bind(template);
+     * String code = template.render(30.0f, 0);
+     * }
+     *
+     * @return The amount of fuel left for nested Template use.
+     */
+    static float fuel() {
+        return Renderer.getCurrent().fuel();
+    }
+
+    /**
+     * Changes the amount of fuel used for the current Template, where the default is
+     * {@link Template#DEFAULT_FUEL_COST}.
+     *
+     * @param fuelCost The amount of fuel used for the current Template.
+     * @return A token for convenient use in {@link Template#body}.
+     */
+    static Token setFuelCost(float fuelCost) {
+        Renderer.getCurrent().setFuelCost(fuelCost);
+        return new NothingToken();
+    }
+
+    /**
+     * Add a {@link Name} in the current scope, i.e. the innermost of either
+     * {@link Template#body} or {@link Hook#set}.
+     * Note that there can be duplicate definitions, and they simply increase
+     * the {@link #weighNames} weight, and increase the probability of sampling
+     * the name with {@link #sampleName}.
+     *
+     * @param name The {@link Name} to be added to the current scope.
+     * @return The token that performs the defining action.
+     */
+    static Token addName(Name name) {
```

Comment:
I have not fully grasped the concept of these `Name`s, yet. The `generateWithNames()` example in `TestTutorial` is quite big and complex. Could you add there a simpler example to better understand this idea of names?

Reply:
Let's discuss offline to see what would be good to have here.